asterius/src/Asterius/CodeGen.hs

```diff
@@ -1465,9 +1466,29 @@ marshalCmmInstr instr = case instr of
 marshalCmmBlockBody :: [GHC.CmmNode GHC.O GHC.O] -> CodeGen [Expression]
 marshalCmmBlockBody instrs = concat <$> for instrs marshalCmmInstr
 
+-- ----------------------------------------------------------------------------
+
+-- | Datatype to capture whether there is at least on CmmSwitch without a
+-- DEFAULT target within a Cmm function. TODO: Place appropriately.
+data ContainsDefault = WithDefault | WithoutDefault
+  deriving (Eq, Show)
+
+instance Monoid ContainsDefault where
+  -- | Always start with the assumption that there are DEFAULTs and everything
+  -- is exhaustive (innocent until proven guilty).
+  mempty = WithDefault
+
+instance Semigroup ContainsDefault where
+  -- | Always start with the assumption that there are DEFAULTs and everything
+  -- is exhaustive (innocent until proven guilty).
+  WithoutDefault <> _ = WithoutDefault
+  WithDefault <> y = y
+
+-- ----------------------------------------------------------------------------
+
 marshalCmmBlockBranch ::
   GHC.CmmNode GHC.O GHC.C ->
-  CodeGen ([Expression], Maybe Expression, [RelooperAddBranch])
+  CodeGen ([Expression], Maybe (Expression, ContainsDefault), [RelooperAddBranch])
```

Comment:
`NeedsUnreachableBlock` should be a standalone field in the outer tuple type. We may produce no `Expression` for a Cmm exit node, but there will still be a `NeedsUnreachableBlock` value (which is false)